File: fakelag-android/app/src/main/java/com/fakelag/android/utils/SettingsManager.kt
package com.fakelag.android.utils

import android.content.Context
import android.content.SharedPreferences

class SettingsManager(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)

    var iconSizeDp: Int
        get() = prefs.getInt(KEY_ICON_SIZE, 58)
        set(value) = prefs.edit().putInt(KEY_ICON_SIZE, value).apply()

    var durasiLagSeconds: Float
        get() = prefs.getFloat(KEY_DURASI_LAG, 1.5f)
        set(value) = prefs.edit().putFloat(KEY_DURASI_LAG, value).apply()

    var durasiTeleSeconds: Float
        get() = prefs.getFloat(KEY_DURASI_TELE, 0.8f)
        set(value) = prefs.edit().putFloat(KEY_DURASI_TELE, value).apply()

    var panelScalePercent: Int
        get() = prefs.getInt(KEY_PANEL_SCALE, 100)
        set(value) = prefs.edit().putInt(KEY_PANEL_SCALE, value).apply()

    var intervalSeconds: Float
        get() = prefs.getFloat(KEY_INTERVAL, 1.0f)
        set(value) = prefs.edit().putFloat(KEY_INTERVAL, value).apply()

    var opacityPercent: Int
        get() = prefs.getInt(KEY_OPACITY, 90)
        set(value) = prefs.edit().putInt(KEY_OPACITY, value).apply()

    var isLockPosition: Boolean
        get() = prefs.getBoolean(KEY_LOCK_POS, false)
        set(value) = prefs.edit().putBoolean(KEY_LOCK_POS, value).apply()

    var isSoundBeep: Boolean
        get() = prefs.getBoolean(KEY_SOUND_BEEP, true)
        set(value) = prefs.edit().putBoolean(KEY_SOUND_BEEP, value).apply()

    var isVolumeControl: Boolean
        get() = prefs.getBoolean(KEY_VOLUME_CTRL, false)
        set(value) = prefs.edit().putBoolean(KEY_VOLUME_CTRL, value).apply()

    var isAutoLag: Boolean
        get() = prefs.getBoolean(KEY_AUTO_LAG, false)
        set(value) = prefs.edit().putBoolean(KEY_AUTO_LAG, value).apply()

    var gameMode: String
        get() = prefs.getString(KEY_GAME_MODE, "FF THƯỜNG") ?: "FF THƯỜNG"
        set(value) = prefs.edit().putString(KEY_GAME_MODE, value).apply()

    var isServiceRunning: Boolean
        get() = prefs.getBoolean(KEY_SERVICE_RUNNING, false)
        set(value) = prefs.edit().putBoolean(KEY_SERVICE_RUNNING, value).apply()

    var lastTeleX: Int
        get() = prefs.getInt(KEY_TELE_X, -1)
        set(value) = prefs.edit().putInt(KEY_TELE_X, value).apply()

    var lastTeleY: Int
        get() = prefs.getInt(KEY_TELE_Y, -1)
        set(value) = prefs.edit().putInt(KEY_TELE_Y, value).apply()

    var lastLagX: Int
        get() = prefs.getInt(KEY_LAG_X, -1)
        set(value) = prefs.edit().putInt(KEY_LAG_X, value).apply()

    var lastLagY: Int
        get() = prefs.getInt(KEY_LAG_Y, -1)
        set(value) = prefs.edit().putInt(KEY_LAG_Y, value).apply()

    var lastPanelX: Int
        get() = prefs.getInt(KEY_PANEL_X, -1)
        set(value) = prefs.edit().putInt(KEY_PANEL_X, value).apply()

    var lastPanelY: Int
        get() = prefs.getInt(KEY_PANEL_Y, -1)
        set(value) = prefs.edit().putInt(KEY_PANEL_Y, value).apply()

    var showTeleButton: Boolean
        get() = prefs.getBoolean(KEY_SHOW_TELE, true)
        set(value) = prefs.edit().putBoolean(KEY_SHOW_TELE, value).apply()

    var showLagButton: Boolean
        get() = prefs.getBoolean(KEY_SHOW_LAG, true)
        set(value) = prefs.edit().putBoolean(KEY_SHOW_LAG, true).apply()

    var savedLicenseKey: String
        get() = prefs.getString(KEY_LICENSE_KEY, "") ?: ""
        set(value) = prefs.edit().putString(KEY_LICENSE_KEY, value).apply()

    var licenseExpiry: String
        get() = prefs.getString(KEY_LICENSE_EXPIRY, "") ?: ""
        set(value) = prefs.edit().putString(KEY_LICENSE_EXPIRY, value).apply()

    var isKeyActivated: Boolean
        get() = prefs.getBoolean(KEY_KEY_ACTIVATED, false)
        set(value) = prefs.edit().putBoolean(KEY_KEY_ACTIVATED, value).apply()

    fun clearLogin() {
        prefs.edit()
            .remove(KEY_LICENSE_KEY)
            .remove(KEY_LICENSE_EXPIRY)
            .putBoolean(KEY_KEY_ACTIVATED, false)
            .apply()
    }

    fun registerListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        prefs.registerOnSharedPreferenceChangeListener(listener)
    }

    fun unregisterListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        prefs.unregisterOnSharedPreferenceChangeListener(listener)
    }

    companion object {
        private const val PREF_NAME = "fakelag_prefs"
        const val KEY_ICON_SIZE = "key_icon_size"
        const val KEY_DURASI_LAG = "key_durasi_lag"
        const val KEY_DURASI_TELE = "key_durasi_tele"
        const val KEY_PANEL_SCALE = "key_panel_scale"
        const val KEY_INTERVAL = "key_interval"
        const val KEY_OPACITY = "key_opacity"
        const val KEY_LOCK_POS = "key_lock_pos"
        const val KEY_SOUND_BEEP = "key_sound_beep"
        const val KEY_VOLUME_CTRL = "key_volume_ctrl"
        const val KEY_AUTO_LAG = "key_auto_lag"
        const val KEY_GAME_MODE = "key_game_mode"
        const val KEY_SERVICE_RUNNING = "key_service_running"
        const val KEY_TELE_X = "key_tele_x"
        const val KEY_TELE_Y = "key_tele_y"
        const val KEY_LAG_X = "key_lag_x"
        const val KEY_LAG_Y = "key_lag_y"
        const val KEY_PANEL_X = "key_panel_x"
        const val KEY_PANEL_Y = "key_panel_y"
        const val KEY_SHOW_TELE = "key_show_tele"
        const val KEY_SHOW_LAG = "key_show_lag"
        const val KEY_LICENSE_KEY = "key_license_key"
        const val KEY_LICENSE_EXPIRY = "key_license_expiry"
        const val KEY_KEY_ACTIVATED = "key_key_activated"

        @Volatile
        private var instance: SettingsManager? = null

        fun getInstance(context: Context): SettingsManager {
            return instance ?: synchronized(this) {
                instance ?: SettingsManager(context).also { instance = it }
            }
        }
    }
}

File: fakelag-ios/FakeLag/Services/SettingsManager.swift
import Foundation
import SwiftUI
import Combine

public final class SettingsManager: ObservableObject {
    public static let shared = SettingsManager()

    private let defaults = UserDefaults.standard

    @Published public var iconSizeDp: CGFloat {
        didSet { defaults.set(Double(iconSizeDp), forKey: "key_icon_size") }
    }

    @Published public var durasiLagSeconds: Double {
        didSet { defaults.set(durasiLagSeconds, forKey: "key_durasi_lag") }
    }

    @Published public var durasiTeleSeconds: Double {
        didSet { defaults.set(durasiTeleSeconds, forKey: "key_durasi_tele") }
    }

    @Published public var panelScalePercent: Int {
        didSet { defaults.set(panelScalePercent, forKey: "key_panel_scale") }
    }

    @Published public var opacityPercent: Int {
        didSet { defaults.set(opacityPercent, forKey: "key_opacity") }
    }

    @Published public var isLockPosition: Bool {
        didSet { defaults.set(isLockPosition, forKey: "key_lock_pos") }
    }

    @Published public var isSoundBeep: Bool {
        didSet { defaults.set(isSoundBeep, forKey: "key_sound_beep") }
    }

    @Published public var gameMode: String {
        didSet { defaults.set(gameMode, forKey: "key_game_mode") }
    }

    @Published public var isServiceRunning: Bool {
        didSet { defaults.set(isServiceRunning, forKey: "key_service_running") }
    }

    @Published public var savedLicenseKey: String {
        didSet { defaults.set(savedLicenseKey, forKey: "key_license_key") }
    }

    @Published public var licenseExpiry: String {
        didSet { defaults.set(licenseExpiry, forKey: "key_license_expiry") }
    }

    @Published public var isKeyActivated: Bool {
        didSet { defaults.set(isKeyActivated, forKey: "key_key_activated") }
    }

    @Published public var showTeleButton: Bool {
        didSet { defaults.set(showTeleButton, forKey: "key_show_tele") }
    }

    @Published public var showLagButton: Bool {
        didSet { defaults.set(showLagButton, forKey: "key_show_lag") }
    }

    @Published public var isPanelShowing: Bool = false

    private init() {
        self.iconSizeDp = CGFloat(defaults.double(forKey: "key_icon_size") != 0 ? defaults.double(forKey: "key_icon_size") : 58.0)
        self.durasiLagSeconds = defaults.double(forKey: "key_durasi_lag") != 0 ? defaults.double(forKey: "key_durasi_lag") : 1.5
        self.durasiTeleSeconds = defaults.double(forKey: "key_durasi_tele") != 0 ? defaults.double(forKey: "key_durasi_tele") : 0.8
        self.panelScalePercent = defaults.integer(forKey: "key_panel_scale") != 0 ? defaults.integer(forKey: "key_panel_scale") : 100
        self.opacityPercent = defaults.integer(forKey: "key_opacity") != 0 ? defaults.integer(forKey: "key_opacity") : 90
        self.isLockPosition = defaults.bool(forKey: "key_lock_pos")
        self.isSoundBeep = defaults.object(forKey: "key_sound_beep") != nil ? defaults.bool(forKey: "key_sound_beep") : true
        self.gameMode = defaults.string(forKey: "key_game_mode") ?? "FF THƯỜNG"
        self.isServiceRunning = defaults.bool(forKey: "key_service_running")
        self.savedLicenseKey = defaults.string(forKey: "key_license_key") ?? ""
        self.licenseExpiry = defaults.string(forKey: "key_license_expiry") ?? ""
        self.isKeyActivated = defaults.bool(forKey: "key_key_activated")
        self.showTeleButton = defaults.object(forKey: "key_show_tele") != nil ? defaults.bool(forKey: "key_show_tele") : true
        self.showLagButton = defaults.object(forKey: "key_show_lag") != nil ? defaults.bool(forKey: "key_show_lag") : true
    }

    public func clearLogin() {
        savedLicenseKey = ""
        licenseExpiry = ""
        isKeyActivated = false
        defaults.removeObject(forKey: "key_license_key")
        defaults.removeObject(forKey: "key_license_expiry")
        defaults.set(false, forKey: "key_key_activated")
    }
}
